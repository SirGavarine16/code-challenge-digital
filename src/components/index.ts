export { default as WelcomeHeader } from './WelcomeHeader';
export { default as SectionLabel } from './SectionLabel';
export { default as PointsDisplay } from './PointsDisplay';
export { default as MovementsList } from './MovementsList';
export { default as MovementCard } from './MovementCard';
export { default as MovementListItemSeparator } from './MovementListItemSeparator';
export { default as Button } from './Button';
export { default as MovementHeader } from './MovementHeader';
export { default as MovementProductImage } from './MovementProductImage';
export { default as EmptyMovementsList } from './EmptyMovementsList';