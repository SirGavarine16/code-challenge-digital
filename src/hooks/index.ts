export { default as useProductsFetch } from './useProductsFetch';
export { default as useProductPoints } from './useProductPoints';