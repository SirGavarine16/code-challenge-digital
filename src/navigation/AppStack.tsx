import { StackNavigationOptions, createStackNavigator } from '@react-navigation/stack';

import { Home, Movement } from '../screens';
import { Movement as MovementData } from '../interfaces';

export type AppStackParams = {
    HomeScreen: undefined;
    MovementScreen: MovementData;
};

const Stack = createStackNavigator<AppStackParams>();

const screenOptions: StackNavigationOptions = {
    headerShown: false,
};

export const AppStack = () => {
    return (
        <Stack.Navigator screenOptions={screenOptions}>
            <Stack.Screen name='HomeScreen' component={Home} />
            <Stack.Screen name='MovementScreen' component={Movement} />
        </Stack.Navigator>
    )
};

export default AppStack;
