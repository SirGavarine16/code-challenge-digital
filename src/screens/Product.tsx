import React, { FC } from 'react';
import { Text, View } from 'react-native';

interface Props {
}

const Product: FC<Props> = () => {
    return (
        <View>
            <Text>Product Screen</Text>
        </View>
    );
};

export default Product;
