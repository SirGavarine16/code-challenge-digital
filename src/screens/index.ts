export { default as Home } from './Home';
export { default as Movement } from './Movement';