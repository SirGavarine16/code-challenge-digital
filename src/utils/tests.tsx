import { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { Movement } from '../interfaces';
import { rootReducer } from '../redux/store';

export const dummyData: Movement = {
    createdAt: "2022-12-09T10:20:00.909Z",
    product: "Fantastic Granite Bacon",
    points: 42416,
    image: "https://loremflickr.com/640/480/technics",
    is_redemption: false,
    id: "3"
}

export const ReduxWrapper = ({ children }: PropsWithChildren<{}>) => {
    const store = configureStore({ reducer: rootReducer });
    return <Provider store={store}>{children}</Provider>
}

export const NavigationWrapper = ({ children }: PropsWithChildren<{}>) => {
    const store = configureStore({ reducer: rootReducer });
    return <NavigationContainer>{children}</NavigationContainer>
}