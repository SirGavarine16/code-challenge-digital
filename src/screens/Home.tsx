import React, { FC } from 'react';
import { StyleSheet, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

import { MovementsList, PointsDisplay, WelcomeHeader } from '../components';
import { useMovementsFetch, useMovementPoints } from '../hooks';
import { spacing } from '../themes/AppTheme';

interface Props {
}

const Home: FC<Props> = () => {
    const { container } = styles;
    const { top: paddingTop, bottom: paddingBottom } = useSafeAreaInsets();

    const { movements } = useMovementsFetch();
    const { totalPoints } = useMovementPoints();

    return (
        <View style={[container, { paddingTop, paddingBottom }]}>
            <WelcomeHeader />
            <PointsDisplay points={totalPoints} />
            <MovementsList bottom={paddingBottom} />
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        paddingHorizontal: spacing.paddingHorizontal,
    }
});

export default Home;
