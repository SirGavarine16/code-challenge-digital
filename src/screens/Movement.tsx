import React, { FC } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { StackScreenProps } from '@react-navigation/stack';

import { Button, MovementHeader, MovementProductImage } from '../components';
import { AppStackParams } from '../navigation/AppStack';
import { colors, spacing } from '../themes/AppTheme';
import { formatDate } from '../utils/formats';

interface Props extends StackScreenProps<AppStackParams, 'MovementScreen'> {
}

const Movement: FC<Props> = ({ route, navigation }) => {
    const { product, image, createdAt, is_redemption, points } = route.params;

    const { top, bottom } = useSafeAreaInsets();
    
    const { container, contentContainer, dataContainer, descriptionLabel, dateLabel, pointsLabel } = styles;

    return (
        <View style={[container, { paddingBottom: bottom === 0 ? 20 : bottom }]}>
            <MovementHeader product={product} top={top} />
            <View style={contentContainer}>
                <MovementProductImage image={image} />
                <View style={dataContainer}>
                    <Text testID='product-details' style={descriptionLabel}>
                        Detalles del producto:
                    </Text>
                    <Text testID='sale-date' style={dateLabel}>
                        {`Comprado el ${formatDate(createdAt)}`}
                    </Text>
                    <Text testID='points-details' style={descriptionLabel}>
                        Con esta compra {is_redemption ? 'canjeaste' : 'acumulaste'}:
                    </Text>
                    <Text testID='points' style={pointsLabel}>
                        {points} puntos
                    </Text>
                </View>
                <Button title='Aceptar' onPress={() => navigation.navigate('HomeScreen')} />
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    contentContainer: {
        flex: 1,
        paddingHorizontal: spacing.paddingHorizontal,
    },
    dataContainer: {
        marginTop: 32,
        flex: 1,
    },
    descriptionLabel: {
        color: colors.textSecondary,
        fontSize: 14,
        fontWeight: '800',
    },
    dateLabel: {
        fontSize: 16,
        fontWeight: '800',
        marginVertical: 20,
    },
    pointsLabel: {
        fontSize: 24,
        fontWeight: '800',
        marginVertical: 32,
    }
})

export default Movement;
