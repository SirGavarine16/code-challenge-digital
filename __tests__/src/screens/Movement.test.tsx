import { render } from '@testing-library/react-native';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

import { Movement } from '../../../src/screens';
import {  dummyData } from '../../../src/utils/tests';
import { AppStackParams } from '../../../src/navigation/AppStack';

describe('<Movement />', () => {

    const renderComponent = () => render(
        <Movement
            route={{ params: dummyData } as RouteProp<AppStackParams, 'MovementScreen'>}
            navigation={{} as StackNavigationProp<AppStackParams, 'MovementScreen'>}
        />
    );

    it('should render correctly', () => {
        expect(renderComponent()).toMatchSnapshot();
    });

    it('should render an accept button', () => {
        const { getByText } = renderComponent();
        expect(getByText('Aceptar')).toBeDefined();
    });

    it('should display the product image', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('card-image')).toBeDefined();
        expect(getByTestId('product-image')).toBeDefined();
    });

    it('should render product details label', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('product-details')).toBeDefined();
    });

    it('should render points details label', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('points-details')).toBeDefined();
    });

    it('should render product details label', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('sale-date')).toBeDefined();
    });

    it('should render points details label', () => {
        const { getByTestId } = renderComponent();
        expect(getByTestId('points')).toBeDefined();
    });
});