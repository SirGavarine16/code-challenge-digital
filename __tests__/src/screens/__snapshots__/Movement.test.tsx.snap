// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movement /> should render correctly 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "paddingBottom": 20,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#CFD6FF",
          "marginBottom": 20,
          "paddingHorizontal": 20,
        },
        {
          "paddingTop": 80,
        },
      ]
    }
  >
    <Text
      style={
        {
          "fontSize": 24,
          "fontWeight": "800",
          "marginBottom": 10,
        }
      }
      testID="product-name"
    >
      Fantastic Granite Bacon
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "aspectRatio": 1,
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 8,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 4.65,
          "width": "100%",
        }
      }
      testID="card-image"
    >
      <Image
        resizeMode="contain"
        source={
          {
            "uri": "https://loremflickr.com/640/480/technics",
          }
        }
        style={
          {
            "backgroundColor": "#F8F8F8",
            "height": "100%",
            "width": "100%",
          }
        }
        testID="product-image"
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "marginTop": 32,
        }
      }
    >
      <Text
        style={
          {
            "color": "#9B9898",
            "fontSize": 14,
            "fontWeight": "800",
          }
        }
        testID="product-details"
      >
        Detalles del producto:
      </Text>
      <Text
        style={
          {
            "fontSize": 16,
            "fontWeight": "800",
            "marginVertical": 20,
          }
        }
        testID="sale-date"
      >
        Comprado el 9 de diciembre, 2022
      </Text>
      <Text
        style={
          {
            "color": "#9B9898",
            "fontSize": 14,
            "fontWeight": "800",
          }
        }
        testID="points-details"
      >
        Con esta compra 
        acumulaste
        :
      </Text>
      <Text
        style={
          {
            "fontSize": 24,
            "fontWeight": "800",
            "marginVertical": 32,
          }
        }
        testID="points"
      >
        42416
         puntos
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "rgb(0, 122, 255)",
          "borderRadius": 10,
          "opacity": 1,
          "padding": 15,
          "width": "100%",
        }
      }
      testID="button"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "800",
            "textAlign": "center",
          }
        }
        testID="button-title"
      >
        Aceptar
      </Text>
    </View>
  </View>
</View>
`;
